[3.14] gh-134698: Hold a lock when the thread state is detached in `ssl` (GH-134724)

gpshead opened the pull request:
Lock when the thread state is detached.
(cherry picked from commit e047a35b23c1aa69ab8d5da56f36319cec4d36b8)


<!-- gh-issue-number: gh-134698 -->
* Issue: gh-134698
<!-- /gh-issue-number -->

ZeroIntensity committed:
[3.14] gh-134698: Hold a lock when the thread state is detached in `ssl` (GH-134724)

Lock when the thread state is detached.
(cherry picked from commit e047a35b23c1aa69ab8d5da56f36319cec4d36b8)

Co-authored-by: Peter Bierma <zintensitydev@gmail.com>
Co-authored-by: Gregory P. Smith <greg@krypto.org>

ZeroIntensity reviewed (approved):
Thanks for doing the backport.

ZeroIntensity committed:
Only lock the SSL context, not the SSL socket.

This solves a deadlock when a socket is blocked while waiting on data,
which ended up causing a major regression in 3.13.6 (see gh-137583).

ZeroIntensity committed:
Merge branch '3.14' into backport-e047a35-3.14
